File: 2022/03/Perl/bin/main.pl
#!perl

use v5.16;
use warnings;
use constant FILENAME => $ENV{ADVENT_INPUT};

my $i = 0;
my %priorities = map { $_ => ++$i } 'a'..'z', 'A'..'Z';

PART1: {
	local @ARGV = FILENAME;
	my $total = 0;
	while ( <> ) {
		chomp;
		my @parts = ( substr( $_, 0, length()/2 ), substr( $_, length()/2 ) );
		my $re = qr/[^$parts[1]]/;
		$total += $priorities{ substr( $parts[0] =~ s/$re//gr, 0, 1 ) };
	}
	say "PART1: $total";
}

PART2: {
	local @ARGV = FILENAME;
	my $total = 0;
	while () {
		chomp( my $e1 = <> // last );
		chomp( my $e2 = <> );
		chomp( my $e3 = <> );
		for my $letter ( 'a'..'z', 'A'..'Z' ) {
			next unless (
				index( $e1, $letter ) >= 0 and
				index( $e2, $letter ) >= 0 and
				index( $e3, $letter ) >= 0
			);
			$total += $priorities{$letter};
		}
	}
	say "PART2: $total";
}

File: 2022/03/Ruby/main.rb
#!/usr/bin/ruby3.0

letters = [*('a'..'z'),*('A'..'Z')]
i = 0
priorities = letters.map { |letter| i += 1; [ letter, i ] }.to_h

total = 0
File.readlines( ENV["ADVENT_INPUT"] ).each do |line|
	line = line.chomp
	first, second = line.chars.each_slice( line.length / 2 ).map( &:join )
	match = first.match /([#{second}])/
	total += priorities[match[1]]
end
puts( "PART1: %s" % [ total ] );

total = 0
File.readlines( ENV["ADVENT_INPUT"] ).each_slice(3) do |line|
	letters.each do |letter|
		if line[0].include? letter and
			line[1].include? letter and
			line[2].include? letter
			total += priorities[letter]
		end
	end
end
puts( "PART2: %s" % [ total ] );
